src/litdata/streaming/dataloader.py

```diff
@@ -700,7 +701,8 @@ def load_state_dict(self, obj: Dict[str, Any]) -> None:
 
         # Inform we are resuming and disable resetting the StreamingDataLoader state.
         # This is toggle back to False when the `__iter__` method of the StreamingDataLoader completes.
-        self.restore = True
+        if obj["num_samples_yielded"] < len(self.dataset):
```

Comment:
This currently forces the dataloader to reset the state if the last epoch was complete, preventing the index error or dataset switching error, but it does solve the source of the index error.

Reply:
After further going through the steps, it seems the index error was caused by different items, but it originated because, in the fresh epoch, it is receiving num_samples_yielded as the full dataset length instead of 0. To fix this, the dataset reset state is required.